<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Interactive Manhattan, Q-Q, and Volcano Plots Using Plotly.js • manhattanly</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Interactive Manhattan, Q-Q, and Volcano Plots Using Plotly.js">
<meta property="og:description" content="manhattanly">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">manhattanly</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/web_only/manhattanly_full.html">Interactive manhattan, volcano and qq plots</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
<li>
  <a href="../articles/manhattanly.html">Get started</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sahirbhatnagar/manhattanly/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="manhattanly_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Interactive Manhattan, Q-Q, and Volcano Plots Using Plotly.js</h1>
                        <h4 class="author">Sahir Bhatnagar</h4>
            
            <h4 class="date">2021-04-22</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sahirbhatnagar/manhattanly/blob/master/vignettes/manhattanly.Rmd"><code>vignettes/manhattanly.Rmd</code></a></small>
      <div class="hidden name"><code>manhattanly.Rmd</code></div>

    </div>

    
    
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{Interactive Manhattan, Q-Q, and Volcano Plots Using Plotly.js}
-->
<p><strong>Author</strong>: <a href="http://sahirbhatnagar.com/">Sahir Bhatnagar</a> (<a href="mailto:sahir.bhatnagar@gmail.com" class="email">sahir.bhatnagar@gmail.com</a>)</p>
<p><strong>Notes</strong>:</p>
<ul>
<li><p>This vignette was built with <code>R markdown</code> and <code>knitr</code>. The source code for this vignette can be found <a href="https://github.com/sahirbhatnagar/manhattanly/blob/master/vignettes/web_only/manhattanly_full.Rmd">here</a>.</p></li>
<li><p>In the interactive manhattan, qq and volcano plots below, we only use a subset (random sample of SNPs on chromosomes 4 to 7) of the <code>HapMap</code> data included in this package. This is to reduce the size of the compiled vignette. To reduce the package size, we don’t output any plots here. Please see <a href="https://sahirbhatnagar.com/manhattanly/articles/web_only/manhattanly_full.html" class="uri">https://sahirbhatnagar.com/manhattanly/articles/web_only/manhattanly_full.html</a> for the full vignette with plots.</p></li>
</ul>
<p><br></p>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p><a href="https://en.wikipedia.org/wiki/Manhattan_plot">Manhattan</a>, <a href="https://en.wikipedia.org/wiki/Q%E2%80%93Q_plot">Q-Q</a> and <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC154570/">volcano</a> plots are popular graphical methods for visualizing results from high-dimensional data analysis such as a (epi)genome wide asssociation study (GWAS or EWAS), in which p-values, Z-scores, test statistics are plotted on a scatter plot against their genomic position. Manhattan plots are used for visualizing potential regions of interest in the genome that are associated with a phenotype. Q-Q plots tell us about the distributional assumptions of the observed test statistics. Volcano plots are the negative log10 p-values plotted against their effect size, odds ratio or log fold-change. They are used to identify clinically meaningful markers in genomic experiments, i.e., markers that are statistically significant and have an effect size greater than some threshold.</p>
<p>Interactive manhattan, Q-Q and volcano plots allow the inspection of specific value (e.g. rs number or gene name) by hovering the mouse over a point, as well as zooming into a region of the genome (e.g. a chromosome) by dragging a rectangle around the relevant area.</p>
<p>This pacakge creates interactive Q-Q, manhattan and volcano plots that are usable from the R console, the <a href="https://www.rstudio.com/"><code>RStudio</code></a> viewer pane, <a href="http://rmarkdown.rstudio.com/"><code>R Markdown</code></a> documents, in <a href="https://plotly.com/dash/"><code>Dash</code></a> apps, <a href="http://shiny.rstudio.com/"><code>Shiny</code></a> apps, embeddable in websites and can be exported as <code>.png</code> files. By hovering the mouse over a point, you can see annotation information such as the SNP identifier and GENE name. You can also drag a rectangle to zoom in on a region of interest and then export the image as a <code>.png</code> file.</p>
<p>This work is based on the <a href="https://github.com/stephenturner/qqman"><code>qqman</code></a> R package and the <a href="https://plot.ly/"><code>plotly.js</code></a> engine. It produces similar manhattan and Q-Q plots as the <code>qqman::manhattan</code> and <code>qqman::qq</code> functions; the main difference here is being able to interact with the plot, including extra annotation information and seamless integration with HTML.</p>
<p><br><br></p>
</div>
<div id="installation" class="section level1">
<h1 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h1>
<p>You can install <code>manhattanly</code> from CRAN:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"manhattanly"</span><span class="op">)</span></code></pre></div>
<p>Alternatively, you can install the development version of <code>manhattanly</code> from <a href="https://github.com/sahirbhatnagar/manhattanly">GitHub</a> with:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">require</a></span><span class="op">(</span><span class="st">"pacman"</span><span class="op">)</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"pacman"</span><span class="op">)</span>
<span class="fu">pacman</span><span class="fu">::</span><span class="fu">p_load_gh</span><span class="op">(</span><span class="st">"sahirbhatnagar/manhattanly"</span><span class="op">)</span></code></pre></div>
<p><br><br></p>
</div>
<div id="quick-start" class="section level1">
<h1 class="hasAnchor">
<a href="#quick-start" class="anchor"></a>Quick Start</h1>
<div id="manhattan-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#manhattan-plot" class="anchor"></a>Manhattan plot</h2>
<p>The <code>manhattanly</code> package ships with an example dataset called <code>HapMap</code>. See <code><a href="../reference/HapMap.html">help(HapMap)</a></code> for more details about how this dataset was created. Here is what the <code>HapMap</code> dataset looks like:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># load the manhattanly library</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/sahirbhatnagar/manhattanly/">manhattanly</a></span><span class="op">)</span></code></pre></div>
<pre><code>## See example usage at http://sahirbhatnagar.com/manhattanly/</code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">12345</span><span class="op">)</span>
<span class="va">HapMap.subset</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html">subset</a></span><span class="op">(</span><span class="va">HapMap</span>, <span class="va">CHR</span> <span class="op">%in%</span> <span class="fl">4</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span>
<span class="co"># for highlighting SNPs of interest</span>
<span class="va">significantSNP</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="va">HapMap.subset</span><span class="op">$</span><span class="va">SNP</span>, <span class="fl">20</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">HapMap.subset</span><span class="op">)</span></code></pre></div>
<pre><code>##      CHR      BP         P        SNP ZSCORE EFFECTSIZE   GENE DISTANCE
## 3300   4  336758 0.7869011  rs6821220 0.2703    -0.1437 ZNF141        0
## 3301   4  992125 0.5116233  rs6855233 0.6563    -0.0128 FGFRL1     3634
## 3302   4 1155741 0.7977199   rs922697 0.2563    -0.1170  SPON2        0
## 3303   4 1302267 0.3590100 rs10025665 0.9173     0.1079   MAEA        0
## 3304   4 1388897 0.1824547 rs11731672 1.3332     0.1005 CRIPAK     9116
## 3305   4 1814818 0.4673265  rs6599401 0.7268     0.0625  LETM1        0</code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">HapMap.subset</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 3410    8</code></pre>
<p>The required columns to create a manhattan plot are the chromosome, base-pair position and p-value. By default, the <code>manhattanly</code> function assumes these columns are named <code>CHR</code>, <code>BP</code> and <code>P</code> (but these can be specified by the user if they are different)</p>
<p>Create an interactive manhattan plot using one command:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/manhattanly.html">manhattanly</a></span><span class="op">(</span><span class="va">HapMap.subset</span>, snp <span class="op">=</span> <span class="st">"SNP"</span>, gene <span class="op">=</span> <span class="st">"GENE"</span><span class="op">)</span></code></pre></div>
<p>The arguments <code>snp = "SNP"</code> and <code>gene = "GENE"</code> specify that we want to add snp and gene information to each point. This information is found in the columns names <code>"SNP"</code> and <code>"GENE"</code> in the <code>HapMap</code> dataset. See <code><a href="../reference/manhattanly.html">help(manhattanly)</a></code> for a full list of options.</p>
</div>
<div id="q-q-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#q-q-plot" class="anchor"></a>Q-Q plot</h2>
<p>Similarly, we can create an interactive Q-Q plot using one command (See <code><a href="../reference/qqly.html">help(qqly)</a></code> for a full list of options):</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/qqly.html">qqly</a></span><span class="op">(</span><span class="va">HapMap.subset</span>, snp <span class="op">=</span> <span class="st">"SNP"</span>, gene <span class="op">=</span> <span class="st">"GENE"</span><span class="op">)</span></code></pre></div>
<p>You can then save the plot as a <code>.png</code> file by clicking on the camera icon in the toolbar (which appears when you hover your mouse over it).</p>
</div>
<div id="volcano-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#volcano-plot" class="anchor"></a>Volcano plot</h2>
<p>You can also make a volcano plot which by default, highlights the points greater than the default <code>genomewideline</code> and <code>effect_size_line</code> arguments:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/volcanoly.html">volcanoly</a></span><span class="op">(</span><span class="va">HapMap.subset</span>, snp <span class="op">=</span> <span class="st">"SNP"</span>, gene <span class="op">=</span> <span class="st">"GENE"</span>, effect_size <span class="op">=</span> <span class="st">"EFFECTSIZE"</span><span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div id="highlighting-snps-of-interest" class="section level2">
<h2 class="hasAnchor">
<a href="#highlighting-snps-of-interest" class="anchor"></a>Highlighting SNPs of Interest</h2>
<p>We can also highlight SNPs of interest using the <code>highlight</code> argument. This package comes with a list of SNPs of interest called <code>significantSNP</code> (see <code><a href="../reference/significantSNP.html">help(significantSNP)</a></code> for more details). To highlight these SNPs we simply pass this vector to the <code>highlight</code> argument (note that these SNPs need to be present in the <code>"SNP"</code> column of your data). We omit the plot here due to size constraints of the package on CRAN:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/manhattanly.html">manhattanly</a></span><span class="op">(</span><span class="va">HapMap.subset</span>, snp <span class="op">=</span> <span class="st">"SNP"</span>, gene <span class="op">=</span> <span class="st">"GENE"</span>, highlight <span class="op">=</span> <span class="va">significantSNP</span><span class="op">)</span></code></pre></div>
</div>
<div id="more-annotations" class="section level2">
<h2 class="hasAnchor">
<a href="#more-annotations" class="anchor"></a>More annotations</h2>
<p>You can add up to 4 annotations. In the following manhattan plot we add the snp, gene, the distance to the nearest gene and the effect size. The same annotations can also be added to Q-Q and volcano plots:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/manhattanly.html">manhattanly</a></span><span class="op">(</span><span class="va">HapMap.subset</span>, snp <span class="op">=</span> <span class="st">"SNP"</span>, gene <span class="op">=</span> <span class="st">"GENE"</span>,
            annotation1 <span class="op">=</span> <span class="st">"DISTANCE"</span>, annotation2 <span class="op">=</span> <span class="st">"EFFECTSIZE"</span>,
            highlight <span class="op">=</span> <span class="va">significantSNP</span><span class="op">)</span></code></pre></div>
<p><br></p>
</div>
<div id="adding-text-annotations-to-the-plot" class="section level2">
<h2 class="hasAnchor">
<a href="#adding-text-annotations-to-the-plot" class="anchor"></a>Adding Text Annotations to the Plot</h2>
<p>The annotations in the previous plots only appear when we hover the mouse over the point. Once we have identified a SNP, or a few SNPs of interest we want to explicitly show the annotation information and save the plot. The output of the <code>manhattanly</code> function is an object which can be further manipulated using the <code><a href="https://magrittr.tidyverse.org/reference/pipe.html">%&gt;%</a></code> operator from the <code>magrittr</code> package:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org">magrittr</a></span><span class="op">)</span>

<span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/manhattanly.html">manhattanly</a></span><span class="op">(</span><span class="va">HapMap.subset</span>, snp <span class="op">=</span> <span class="st">"SNP"</span>, gene <span class="op">=</span> <span class="st">"GENE"</span>,
            annotation1 <span class="op">=</span> <span class="st">"DISTANCE"</span>, annotation2 <span class="op">=</span> <span class="st">"EFFECTSIZE"</span>,
            highlight <span class="op">=</span> <span class="va">significantSNP</span><span class="op">)</span>

<span class="co"># get the x and y coordinates from the pre-processed data</span>
<span class="va">plotData</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/manhattanr.html">manhattanr</a></span><span class="op">(</span><span class="va">HapMap.subset</span>, snp <span class="op">=</span> <span class="st">"SNP"</span>, gene <span class="op">=</span> <span class="st">"GENE"</span><span class="op">)</span><span class="op">[[</span><span class="st">"data"</span><span class="op">]</span><span class="op">]</span>

<span class="co"># annotate the smallest p-value</span>
<span class="va">annotate</span> <span class="op">&lt;-</span> <span class="va">plotData</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html">which.min</a></span><span class="op">(</span><span class="va">plotData</span><span class="op">$</span><span class="va">P</span><span class="op">)</span>,<span class="op">]</span>

<span class="co"># x and y coordinates of SNP with smallest p-value</span>
<span class="va">xc</span> <span class="op">&lt;-</span> <span class="va">annotate</span><span class="op">$</span><span class="va">pos</span>
<span class="va">yc</span> <span class="op">&lt;-</span> <span class="va">annotate</span><span class="op">$</span><span class="va">logp</span>

<span class="va">p</span> <span class="op">%&gt;%</span> <span class="fu">plotly</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/plotly/reference/layout.html">layout</a></span><span class="op">(</span>annotations <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
  <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">xc</span>, y <span class="op">=</span> <span class="va">yc</span>,
       text <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span><span class="op">(</span><span class="va">annotate</span><span class="op">$</span><span class="va">SNP</span>,<span class="st">"&lt;br&gt;"</span>,<span class="st">"GENE: "</span>,<span class="va">annotate</span><span class="op">$</span><span class="va">GENE</span><span class="op">)</span>,
       font <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>family <span class="op">=</span> <span class="st">"serif"</span>, size <span class="op">=</span> <span class="fl">10</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p>You can then save the plot as a <code>.png</code> file by clicking on the camera icon in the toolbar (which appears when you hover your mouse over it).</p>
<p><br></p>
</div>
<div id="sharing-your-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#sharing-your-plots" class="anchor"></a>Sharing your Plots</h2>
<p>By default, plotly for R runs locally in your web browser or in the R Studio viewer. It would be useful to be able to easily share these plot with for example, your collaborators or supervisor, especially in the during the exploratory data analysis stage of your project.</p>
<p>You can publish your charts to the web with <a href="https://plot.ly/">plotly’s web service</a> in three steps.</p>
<div id="step-1-signup-for-a-free-plotly-account" class="section level3">
<h3 class="hasAnchor">
<a href="#step-1-signup-for-a-free-plotly-account" class="anchor"></a>Step 1: Signup for a free plotly account</h3>
<p>Create a free plotly account <a href="https://plot.ly/">here</a>. A plotly account is required to publish charts online. It’s free to get started, and you control the privacy of your charts.</p>
</div>
<div id="step-2-save-your-authentication-credentials" class="section level3">
<h3 class="hasAnchor">
<a href="#step-2-save-your-authentication-credentials" class="anchor"></a>Step 2: Save your authentication credentials</h3>
<p>Find your authentication API keys in your online settings. Set them in your R session with:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html">Sys.setenv</a></span><span class="op">(</span><span class="st">"plotly_username"</span><span class="op">=</span><span class="st">"your_plotly_username"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Sys.setenv.html">Sys.setenv</a></span><span class="op">(</span><span class="st">"plotly_api_key"</span><span class="op">=</span><span class="st">"your_api_key"</span><span class="op">)</span></code></pre></div>
</div>
<div id="step-3-publish-your-graphs-to-plotly-with-plotlyapi_create" class="section level3">
<h3 class="hasAnchor">
<a href="#step-3-publish-your-graphs-to-plotly-with-plotlyapi_create" class="anchor"></a>Step 3: Publish your graphs to plotly with <code>plotly::api_create()</code>
</h3>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://plotly-r.com">plotly</a></span><span class="op">)</span>
<span class="co"># p is the interactive manhattan plot we saved earlier</span>
<span class="fu">plotly</span><span class="fu">::</span><span class="fu"><a href="https://docs.ropensci.org/plotly/reference/api.html">api_create</a></span><span class="op">(</span><span class="va">p</span>, filename <span class="op">=</span> <span class="st">"r-docs/manhattan"</span>, world_readable<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<ul>
<li>
<code>filename</code> sets the name of the file inside your online plotly account.</li>
<li>
<code>world_readable</code> sets the privacy of your chart. If <code>TRUE</code>, the graph is publically viewable, if <code>FALSE</code>, only you can view it.</li>
</ul>
<p><br><br></p>
</div>
</div>
</div>
<div id="dynamic-documents-with-knitr-and-r-markdown" class="section level1">
<h1 class="hasAnchor">
<a href="#dynamic-documents-with-knitr-and-r-markdown" class="anchor"></a>Dynamic Documents with <code>knitr</code> and <code>R Markdown</code>
</h1>
<p><a href="http://rmarkdown.rstudio.com/"><code>R Markdown</code></a> is a an <code>R</code> software package that allows the creation of dynamic documents, i.e., embed <code>R</code> code with text to create fully reproducible reports. Furthermore it allows easy creation of <code>HTML</code> reports without knowing how to code in <code>HTML</code> (such as this vignette). This means you can embed interactive manhattan, qq and volcano plots in <code>HTML</code> reports using the <code>manhattanly</code> package. For example, to embed the above manhattan plot I included the following code chunk in the <a href="https://raw.githubusercontent.com/sahirbhatnagar/manhattanly/master/vignettes/web-only/manhattanly_full.Rmd"><code>.Rmd</code></a> document:</p>
<pre><code>```{r}
library(plotly)
manhattanly(subset(HapMap, CHR %in% 4:7), snp = "SNP", gene = "GENE")
```</code></pre>
</div>
<div id="details" class="section level1">
<h1 class="hasAnchor">
<a href="#details" class="anchor"></a>Details</h1>
<div id="data-pre-processing" class="section level2">
<h2 class="hasAnchor">
<a href="#data-pre-processing" class="anchor"></a>Data Pre-Processing</h2>
<p>The <code>manhattanly</code> package splits up the data pre-processing from the rendering of the plot object (inspired by the <a href="https://github.com/talgalili/heatmaply"><code>heatmaply</code></a> package. These are done by the <code>manhattanr</code>, <code>qqr</code> and <code>volcanor</code> functions:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create an object of class `manhattanr`</span>
<span class="va">manhattanrObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/manhattanr.html">manhattanr</a></span><span class="op">(</span><span class="va">HapMap</span><span class="op">)</span>

<span class="co"># whats in there</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">manhattanrObject</span><span class="op">)</span></code></pre></div>
<pre><code>## List of 10
##  $ data           :'data.frame': 14412 obs. of  6 variables:
##   ..$ CHR  : int [1:14412] 1 1 1 1 1 1 1 1 1 1 ...
##   ..$ BP   : int [1:14412] 937641 1136887 2116240 2310562 2681715 2917484 2942700 3298358 3501155 3676178 ...
##   ..$ P    : num [1:14412] 0.335 0.246 0.823 0.493 0.605 ...
##   ..$ logp : num [1:14412] 0.4745 0.6093 0.0844 0.307 0.218 ...
##   ..$ pos  : num [1:14412] 937641 1136887 2116240 2310562 2681715 ...
##   ..$ index: num [1:14412] 1 1 1 1 1 1 1 1 1 1 ...
##  $ xlabel         : chr "Chromosome"
##  $ ticks          : num [1:23] 1.24e+08 3.68e+08 5.88e+08 7.83e+08 9.68e+08 ...
##  $ labs           : int [1:23] 1 2 3 4 5 6 7 8 9 10 ...
##  $ nchr           : int 23
##  $ pName          : chr "P"
##  $ snpName        : logi NA
##  $ geneName       : logi NA
##  $ annotation1Name: logi NA
##  $ annotation2Name: logi NA
##  - attr(*, "class")= chr "manhattanr"</code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># the data used for plotting is a data.frame</span>
<span class="co"># this data.frame can be used with any graphics function such as in base R</span>
<span class="co"># you do not need to use plotly</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">manhattanrObject</span><span class="op">[[</span><span class="st">"data"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##   CHR      BP         P       logp     pos index
## 1   1  937641 0.3353438 0.47450973  937641     1
## 2   1 1136887 0.2458571 0.60931719 1136887     1
## 3   1 2116240 0.8232859 0.08444933 2116240     1
## 4   1 2310562 0.4932038 0.30697357 2310562     1
## 5   1 2681715 0.6053916 0.21796358 2681715     1
## 6   1 2917484 0.1944431 0.71120743 2917484     1</code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">is.data.frame</a></span><span class="op">(</span><span class="va">manhattanrObject</span><span class="op">[[</span><span class="st">"data"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
<p>This <code>manhattanrObject</code> which is of class <code>manhattanr</code> can also be passed to the <code>manhattanly</code> function:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/manhattanly.html">manhattanly</a></span><span class="op">(</span><span class="va">manhattanrObject</span><span class="op">)</span></code></pre></div>
</div>
<div id="more-annotations-1" class="section level2">
<h2 class="hasAnchor">
<a href="#more-annotations-1" class="anchor"></a>More annotations</h2>
<p>We can specify more annotations in the data using the <code>snp</code>, <code>gene</code>, <code>annotation1</code> and <code>annotation2</code> arguments:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># create an object of class `manhattanr`</span>
<span class="va">manhattanrObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/manhattanr.html">manhattanr</a></span><span class="op">(</span><span class="va">HapMap</span>, snp <span class="op">=</span> <span class="st">"SNP"</span>, gene <span class="op">=</span> <span class="st">"GENE"</span>,
            annotation1 <span class="op">=</span> <span class="st">"DISTANCE"</span>, annotation2 <span class="op">=</span> <span class="st">"EFFECTSIZE"</span><span class="op">)</span>

<span class="co"># the annotation columns are now part of the data.frame</span>
<span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">manhattanrObject</span><span class="op">[[</span><span class="st">"data"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##   CHR      BP         P        SNP    GENE DISTANCE EFFECTSIZE       logp
## 1   1  937641 0.3353438  rs9697358   ISG15     1068    -0.0946 0.47450973
## 2   1 1136887 0.2458571 rs34945898 TNFRSF4        0    -0.0947 0.60931719
## 3   1 2116240 0.8232859 rs12034613  FP7162        0    -0.0741 0.08444933
## 4   1 2310562 0.4932038  rs4648633   MORN1        0     0.0146 0.30697357
## 5   1 2681715 0.6053916  rs4430271   MMEL1   127427     0.1234 0.21796358
## 6   1 2917484 0.1944431  rs6685625  ACTRT2    10421     0.1979 0.71120743
##       pos index
## 1  937641     1
## 2 1136887     1
## 3 2116240     1
## 4 2310562     1
## 5 2681715     1
## 6 2917484     1</code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html">is.data.frame</a></span><span class="op">(</span><span class="va">manhattanrObject</span><span class="op">[[</span><span class="st">"data"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] TRUE</code></pre>
</div>
<div id="q-q-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#q-q-plots" class="anchor"></a>Q-Q plots</h2>
<p>Similarly the data used for the Q-Q plot can be created using the <code>qqr</code> function:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">qqrObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/qqr.html">qqr</a></span><span class="op">(</span><span class="va">HapMap</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">qqrObject</span><span class="op">)</span></code></pre></div>
<pre><code>## List of 6
##  $ data           :'data.frame': 14412 obs. of  3 variables:
##   ..$ P       : num [1:14412] 6.75e-10 3.41e-09 3.95e-09 4.71e-09 5.02e-09 ...
##   ..$ OBSERVED: num [1:14412] 9.17 8.47 8.4 8.33 8.3 ...
##   ..$ EXPECTED: num [1:14412] 4.46 3.98 3.76 3.61 3.51 ...
##  $ pName          : chr "P"
##  $ snpName        : logi NA
##  $ geneName       : logi NA
##  $ annotation1Name: logi NA
##  $ annotation2Name: logi NA
##  - attr(*, "class")= chr "qqr"</code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">qqrObject</span><span class="op">[[</span><span class="st">"data"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##                P OBSERVED EXPECTED
## 4346 6.75010e-10 9.170690 4.459754
## 4347 3.41101e-09 8.467117 3.982633
## 4344 3.95101e-09 8.403292 3.760784
## 4338 4.70701e-09 8.327255 3.614656
## 4342 5.02201e-09 8.299122 3.505512
## 4341 6.22801e-09 8.205651 3.418362</code></pre>
<p>This <code>qqrObject</code> which is of class <code>qqr</code> can also be passed to the <code>qqly</code> function (we omit the plot here for the sake of size of the rendered vignette):</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/qqly.html">qqly</a></span><span class="op">(</span><span class="va">qqrObject</span><span class="op">)</span></code></pre></div>
</div>
<div id="volcano-plots" class="section level2">
<h2 class="hasAnchor">
<a href="#volcano-plots" class="anchor"></a>Volcano plots</h2>
<p>Similarly the data used for the volcanor plot can be created using the <code>volcanor</code> function:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">volcanorObject</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/volcanor.html">volcanor</a></span><span class="op">(</span><span class="va">HapMap</span>, 
                        p <span class="op">=</span> <span class="st">"P"</span>,
                        effect_size <span class="op">=</span> <span class="st">"EFFECTSIZE"</span>,
                        snp <span class="op">=</span> <span class="st">"SNP"</span>,
                        gene <span class="op">=</span> <span class="st">"GENE"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">volcanorObject</span><span class="op">)</span></code></pre></div>
<pre><code>## List of 8
##  $ data           :'data.frame': 14412 obs. of  5 variables:
##   ..$ EFFECTSIZE: num [1:14412] -0.0946 -0.0947 -0.0741 0.0146 0.1234 ...
##   ..$ P         : num [1:14412] 0.335 0.246 0.823 0.493 0.605 ...
##   ..$ SNP       : chr [1:14412] "rs9697358" "rs34945898" "rs12034613" "rs4648633" ...
##   ..$ GENE      : chr [1:14412] "ISG15" "TNFRSF4" "FP7162" "MORN1" ...
##   ..$ LOG10P    : num [1:14412] 0.4745 0.6093 0.0844 0.307 0.218 ...
##  $ pName          : chr "P"
##  $ effectName     : chr "EFFECTSIZE"
##  $ xlabel         : chr "EFFECTSIZE"
##  $ snpName        : chr "SNP"
##  $ geneName       : chr "GENE"
##  $ annotation1Name: logi NA
##  $ annotation2Name: logi NA
##  - attr(*, "class")= chr "volcanor"</code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">volcanorObject</span><span class="op">[[</span><span class="st">"data"</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></code></pre></div>
<pre><code>##   EFFECTSIZE         P        SNP    GENE     LOG10P
## 1    -0.0946 0.3353438  rs9697358   ISG15 0.47450973
## 2    -0.0947 0.2458571 rs34945898 TNFRSF4 0.60931719
## 3    -0.0741 0.8232859 rs12034613  FP7162 0.08444933
## 4     0.0146 0.4932038  rs4648633   MORN1 0.30697357
## 5     0.1234 0.6053916  rs4430271   MMEL1 0.21796358
## 6     0.1979 0.1944431  rs6685625  ACTRT2 0.71120743</code></pre>
<p>This can be directly passed to the <code>volcanoly</code> function, with additional arguments:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/volcanoly.html">volcanoly</a></span><span class="op">(</span><span class="va">volcanorObject</span>, effect_size_line <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://sahirbhatnagar.com/">Sahir Bhatnagar</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
